pandas/tests/indexes/multi/test_integrity.py

```diff
@@ -254,7 +254,7 @@ def test_rangeindex_fallback_coercion_bug():
 def test_hash_error(indices):
     index = indices
     with pytest.raises(
-        TypeError, match=("unhashable type: {0.__name__!r}".format(type(index)))
+        TypeError, match=(f"unhashable type: {repr(type(index).__name__)}")
```

Comment:
so for something we *know* is a string (e.g. ``__name__``) then I think we could dispense with the ``repr``, however if its not then ok with repr (but again its going to get coerced with ``str()`` anyhow, so it will mostly wok).

So I think we should actually limit our use of repr, and esp NOT when its a string (e.g. ``__name__``)

Reply:
@jreback please correct if me if I'm wrong, in simple English you mean to not use the ```repr(foo)``` when for example ```foo``` or one of its attributes is obviously a string? for example ```foo.__name__```?

Reply:
yep that's exactly write. I think we can definitely eliminate those cases (so can add a rule eventually for that too)

Reply:
We should really concentrate all of these rules in the [pandas style guidelines](https://dev.pandas.io/docs/development/contributing.html#style-guidelines).

I think it should be considered adding a ```Unwanted patterns``` subsection, describing  all the unwanted patterns. for e.g

```foo.__class__.__name``` --> ```type(foo).__name__```

And

```{foo!r}``` --> ```{repr(foo)}``` # Unless ```foo``` is ...

@jreback do you think it's a good idea?

Reply:
yes an Unwanted patterns section is good (we already have a number of these, they are in other sections now); certainly this would be a followup